DOC: Improve SIMD documentation(1/4)

seiko2plus opened the pull request:
related to #18926, merge after #18884

This pull request is limited to improve the following : 
- [x] Features generator
- [x] Split into multiple pages
~- [ ] CPU build options(remains runtime trace)~
~- [ ] How to use the dispatcher API~
~- [ ] How the dispatcher works~

This pull request will not cover the following:
- Universal intrinsics API (Going to be implemented with C++ interface  in the next pr)
- How to extend the infrastructure with more CPU features and architectures (requires improve the docstring of CCompilerOpt)

----

Live URL https://23836-908607-gh.circle-artifacts.com/0/doc/build/html/reference/simd/index.html

seiko2plus committed:
DOC, SIMD: Improve the auto-generated tables of CPU features

seiko2plus committed:
DOC, SIMD: add a new index for the optimization page to separate into multiple files

seiko2plus committed:
DOC, SIMD: Improve build options and move them into a separated page

seiko2plus committed:
DOC, SIMD: move How It Works into separated page

rgommers commented:
@seiko2plus this is still draft, however there is valuable content in here already. Should we perhaps review and merge this as is, and leave the TODO items for a next PR?

seiko2plus commented on doc/source/reference/simd/simd-optimizations.rst:
```diff
@@ -1,527 +1,3 @@
-
-
-.. _`Universal Intrinsics`: https://numpy.org/neps/nep-0038-SIMD-optimizations.html
+TODO add redirect
```
How to add a redirect 301 to the new index?

charris replied:
@melissawm Thoughts?

melissawm replied:
I've never done this before, but I found this extension for redirects in Sphinx: https://github.com/sphinx-contrib/redirects

Not sure if there are easier solutions... I'll do some more research.

melissawm replied:
Also this package: https://pypi.org/project/sphinx-reredirects/

seiko2plus replied:
it's not possible to add redirect http/301(SEO friendly), since the documents served as the static pages and unless if github(host) provides a way for manually routing rules. anyway, I added just a normal meta refresh for the new destination.

melissawm replied:
Sounds good, thanks!

seiko2plus commented:
@rgommers, sounds good to me,  for the next pr then.

melissawm commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+
+**Platform**:
+
+- :enabled:`Architecture`: The architecture name of target CPU, it should be one of
+  (x86, x64, ppc64, ppc64le, armhf, aarch64, unknown)
```
Very nit, but I'd add some formatting to make sure this is visibly a series of keywords:
```suggestion
- :enabled:`Architecture`: The architecture name of target CPU. It should be one of
  ``x86``, ``x64``, ``ppc64``, ``ppc64le``, ``armhf``, ``aarch64`` or ``unknown``.
```

melissawm commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+- :enabled:`Architecture`: The architecture name of target CPU, it should be one of
+  (x86, x64, ppc64, ppc64le, armhf, aarch64, unknown)
+
+- :enabled:`Compiler`: The compiler name, it should be one of
+  (gcc, clang, msvc, icc, iccw, unix-like)
```
```suggestion
- :enabled:`Compiler`: The compiler name. It should be one of
  gcc, clang, msvc, icc, iccw or unix-like.
```

melissawm commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+
+- :enabled:`Requested`: The specific features and options to ``--cpu-baseline`` as-is.
+- :enabled:`Enabled`: The final set of enabled CPU features.
+- :enabled:`Flags`: The compiler flags that been used to all NumPy `C/C++` sources
+  during the compilation except for temporary sources that been used for generating
+  the binary objects of dispatched features.
```
```suggestion
- :enabled:`Flags`: The compiler flags that were used to all NumPy `C/C++` sources
  during the compilation except for temporary sources that have been used for generating
  the binary objects of dispatched features.
```

melissawm commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+  the features for which optimizations have been generated are shown in the
+  form of several sections with similar properties explained as follows:
+
+    - :enabled:`One or multiple dispatched feature`: The implied CPU features.
+    - :enabled:`Flags`: The compiler flags that been used for these features.
+    - :enabled:`Extra checks`: Similar to the baseline but for these dispatched features.
+    - :enabled:`Detect`: Set of CPU features that need be detected in runtime in order to
+      execute the generated optimizations.
+    - The lines that come after the above property and end with a ':' on a separate line,
+      represent the paths of c/c++ sources that define the generated optimizations.
```
```suggestion
  - :enabled:`One or multiple dispatched feature`: The implied CPU features.
  - :enabled:`Flags`: The compiler flags that been used for these features.
  - :enabled:`Extra checks`: Similar to the baseline but for these dispatched features.
  - :enabled:`Detect`: Set of CPU features that need be detected in runtime in order to
    execute the generated optimizations.
  - The lines that come after the above property and end with a ':' on a separate line,
    represent the paths of c/c++ sources that define the generated optimizations.
```

melissawm commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+
+Runtime Trace
+-------------
+TODO
```
Maybe we can remove the TODO and write something like "To be completed"?

melissawm commented on doc/source/reference/simd/how-it-works.rst:
```diff
@@ -0,0 +1,349 @@
+*****************************
+How the CPU dispatcher works?
```
```suggestion
How does the CPU dispatcher work?
```

melissawm commented on doc/source/reference/simd/how-it-works.rst:
```diff
@@ -0,0 +1,349 @@
+How the CPU dispatcher works?
+*****************************
+
+NumPy dispatcher is based on multi-source compiling, which means taking
+a certain source and compiling it multiple times with different compiler
+flags and also with different **C** definitions that affect the code
+paths to enable certain instruction-sets for each compiled object
+depending on the required optimizations, then combining the returned
+objects together.
```
This is a very long sentence - might be re-worded, maybe something like:
```suggestion
NumPy dispatcher is based on multi-source compiling, which means taking
a certain source and compiling it multiple times with different compiler
flags and also with different **C** definitions that affect the code
paths. This enables certain instruction-sets for each compiled object
depending on the required optimizations which are then combined in the returned
objects together.
```

seiko2plus replied:
Done, but I have changed the last phrase to become "and ends with linking the
returned objects together.".

melissawm commented on doc/source/reference/simd/how-it-works.rst:
```diff
@@ -0,0 +1,349 @@
+.. figure:: ../figures/opt-infra.png
+
+This mechanism should support all compilers and it doesn't require any
+compiler-specific extension, but at the same time it is adds a few steps to
+normal compilation that are explained as follows:
```
```suggestion
compiler-specific extension, but at the same time it adds a few steps to
normal compilation that are explained as follows.
```

melissawm commented on doc/source/reference/simd/how-it-works.rst:
```diff
@@ -0,0 +1,349 @@
+validated during the previous step.
+
+It also contains extra C definitions that are used for defining NumPy's
+Python-level module attributes ``__cpu_baseline__`` and ``__cpu_dispaٍtch__``.
```
```suggestion
Python-level module attributes ``__cpu_baseline__`` and ``__cpu_dispatch__``.
```

melissawm commented on doc/source/reference/simd/how-it-works.rst:
```diff
@@ -0,0 +1,349 @@
+
+  The generated header takes the name of the dispatch-able source after
+  excluding the extension and replace it with '**.h**', for example
+  assume we have a dispatch-able source called **hello.dispatch.c** and
```
Because you were using double backticks for filenames before, I think it would be nice to keep a consistent formatting and use double backticks for all filenames.

melissawm reviewed (commented):
Hello @seiko2plus ! I did a thorough read and this looks really good. I left a number of comments, most of them about grammar, typos or formatting, nothing major. One point I would like to raise is that we could move this document to the "Under the hood" section, as I think it fits better there. But I'll leave the decision to you and other maintainers. Also feel free to ignore my nitpicky comments!

seiko2plus commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+- ``NONE``: enable no features.
+
+- ``NATIVE``: Enables all CPU features that supported by the host CPU,
+  this operation is based on the compiler flags (`-march=native, -xHost, /QxHost`)
```
I think it's better to wrap each flag instead, and yes it's really `QxHost` see [
Intel oneAPI guide](https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/code-generation-options/xhost-qxhost.html) for more details.

seiko2plus committed:
DOC, SIMD: Apply suggestions from Melissa's code review

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

seiko2plus committed:
DOC, SIMD: redirect the old path to the new index

seiko2plus commented on doc/source/reference/simd/index.rst:
```diff
@@ -0,0 +1,42 @@
+
+The optimization process in NumPy is carried out in three layers:
+
+- Code is *written* using the universal intrinsics, with guards that
```
add a simple explanation for it but I am not sure if that would be enough. we will have a separate document for it with a reference for the supported intrinsics.

seiko2plus commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+
+.. note::
+
+    If `build_clib`` or `build_ext` are not specified by the user,
```
done by backticks.

seiko2plus commented on doc/source/reference/simd/build-options.rst:
```diff
@@ -0,0 +1,373 @@
+      that are not available in the target CPU.
+
+These options are accessible through Distutils commands
+`build`, `build_clib` and `build_ext`, they accept a set of
```
done as you suggested.

seiko2plus commented:
Hi @melissawm!, thank you for your amazing and accurate review. I have addressed all your points within a separate commit. Now what is left is moving this index to "Under the hood", I'm not sure, to be honest of what the right call should be but the "build options" document shouldn't be part of the "under the hood" since its targets the end-user but it makes sense to move the rest and the upcoming documents to "under the hood". I'd rather leave that decision up to you and the doc team.

rgommers commented on doc/source/reference/simd/index.rst:
```diff
@@ -0,0 +1,43 @@
+    build-options
+    how-it-works
+
+.. _`NEP-38`: https://numpy.org/neps/nep-0038-SIMD-optimizations.html
```
This isn't needed. We have intersphinx set up, so simply this should do it:
```
`NEP 38`_
```
The reference 6 lines up just needs the dash removed.

rgommers replied:
This is a minor thing, so let's ignore it - maybe you can take it along in a future PR.

rgommers reviewed (approved):
I had a read through, this looks great. The features generator is a good idea, keeps the detailed listing of supported features maintainable. Thanks @seiko2plus, and thanks @melissawm for the detailed review!
